ui/de-lib/src/main/java/org/iplantc/de/admin/desktop/client/toolAdmin/service/impl/ToolAdminServiceFacadeImpl.java

```diff
@@ -0,0 +1,111 @@
+    @Inject private DiscEnvApiService deService;
+
+    @Override
+    public void getTools(String searchTerm, AsyncCallback<List<Tool>> callback) {
+        String address = TOOLS + "?search=" + URL.encodeQueryString(searchTerm);
+
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(GET, address);
+        deService.getServiceData(wrapper, new ToolListCallbackConverter(callback, factory));
+    }
+
+    @Override
+    public void getToolDetails(String toolId, AsyncCallback<Tool> callback) {
+        String address = TOOLS + "/" + toolId;
+
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(GET, address);
+        deService.getServiceData(wrapper, new ToolCallbackConverter(callback, factory));
+    }
+
+    @Override
+    public void addTool(ToolList toolList, AsyncCallback<Void> callback) {
+        String address = TOOLS_ADMIN;
+
+        final Splittable encode = AutoBeanCodex.encode(AutoBeanUtils.getAutoBean(toolList));
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(POST, address, encode.getPayload());
+        deService.getServiceData(wrapper, new StringToVoidCallbackConverter(callback));
+    }
+
+    @Override
+    public void updateTool(Tool tool, boolean overwrite, AsyncCallback<Void> callback) {
+
+        String toolId = tool.getId();
+        String address;
+        if (overwrite) {
+            address = TOOLS_ADMIN + "/" + toolId + "?overwrite-public=true";
+        } else {
+            address = TOOLS_ADMIN + "/" + toolId;
+        }
+
+        final Splittable encode = AutoBeanCodex.encode(AutoBeanUtils.getAutoBean(tool));
+        String payload = removeEmptyValues(encode);
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(PATCH, address, payload);
+        deService.getServiceData(wrapper, new StringToVoidCallbackConverter(callback));
+    }
+
+    @Override
+    public void deleteTool(String toolId, AsyncCallback<Void> callback) {
+        String address = TOOLS_ADMIN + "/" + toolId;
+
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(DELETE, address);
+        deService.getServiceData(wrapper, new StringToVoidCallbackConverter(callback));
+    }
+
+    private String removeEmptyValues(Splittable payload) {
+        removeEmptyValuesHelper(payload);
+        return payload.getPayload();
+    }
+
+    private void removeEmptyValuesHelper(Splittable obj) {
+        List<String> propertyKeys = obj.getPropertyKeys();
+
+        for (String key : propertyKeys) {
+            Splittable value = obj.get(key);
+            String temp = value.getPayload().replaceAll("[\\[\\]\\{\\}]", "");
+            if ((temp.equals("") || temp.equals("\"\"")) && !key.matches("(input|output)_files")) {
+                StringQuoter.nullValue().assign(obj, key);
+            } else {
+                if (value.isIndexed() || value.isKeyed()) {
+                    removeEmptyValuesHelper(value);
+                }
+            }
+        }
+    }
+}
```

Comment:
A `ValueChangeHandler` could set a null value instead of an empty string for a `TextField`. For example:
```java
        workingDirectoryEditor.addValueChangeHandler(new ValueChangeHandler<String>() {
            @Override
            public void onValueChange(ValueChangeEvent<String> valueChangeEvent) {
                final String value = valueChangeEvent.getValue();
                if (value != null && value.isEmpty()) {
                    workingDirectoryEditor.setValue(null);
                }
            }
        });
```
This `ValueChangeHandler` could be extracted into its own class and reused in every `TextField` we don't want submitting empty-strings.

I'm not sure about the list editors, but the service endpoints shouldn't store empty values for the `container_devices`, `container_volumes`, and `container_volumes_from` arrays if they are just empty arrays, as long as we don't submit them with empty strings in their elements. So we don't have to null-out these arrays for now.